### PR TITLE
Fix for #4520 '-ArgumentList should accept @() or $null'

### LINKED ISSUE #4520: Start-Process -ArgumentList should accept an empty array / the null collection and interpret that as "no arguments to pass"
`-ArgumentList` (`-Args`) currently requires that a _non-empty_ array be passed, which is inconvenient in scenarios where you want to pass arguments through, whether any were specified or not, such as via `$Args`.

An empty array or a null collection (`[System.Management.Automation.Internal.AutomationNull]::Value`) should be accepted and should be interpreted as "no arguments to pass".

Currently, passing `$Args` through in a function is cumbersome; imagine a function `wpsh` whose purpose is to invoke a Windows PowerShell window from PowerShell Core, with any arguments passed through:

```powershell
function wpsh {
  # Pass all arguments through. 
  # To do this in a single Start-Process call, you need a cumbersome workaround
  # via an aux. hashtable and splatting.
  # Ideally,  
  #    Start-Process (Get-Command powershell.exe).Path -Args $Args 
  # should work directly.
  $htArgs  = if ($Args.Count) { @{ Args = $Args } } else { @{} }
  Start-Process (Get-Command powershell.exe).Path @htArgs
}
```

Steps to reproduce
------------------

```powershell
# Or: [System.Management.Automation.Internal.AutomationNull]::Value, such as produced by `& {}`
$arguments = @()  
Start-Process code -Args $arguments
```

Expected behavior
-----------------

`code` (the Visual Studio CLI, if installed) should open / activate, with no arguments being passed.

Actual behavior
---------------

Invocation fails with the following error message:

```none
Start-Process : Cannot validate argument on parameter 'ArgumentList'.
The argument is null, empty, or an element of the argument collection contains a null value.
Supply a  collection that does not contain any null values and then try the command again.
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.5 on macOS 10.12.6
PowerShell Core v6.0.0-beta.5 on Ubuntu 16.04.3 LTS
PowerShell Core v6.0.0-beta.5 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.483 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

### PR DESCRIPTION
**Pull request by BrucePay**
Fix #4520 

Just removed the [ValidateIsNotNulOrEmpty] attribute from the ArgumentList parameter. (Also removed some obsolete comments).

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 1]
```diff
@@ -122,7 +122,19 @@ Describe "Start-Process" -Tags @("Feature") {
...
+    }
+
+   It "Should run without errors when -ArgumentList is null" {
+        $process = Start-Process $pingCommand -ArgumentList $null -PassThru -WindowStyle Hidden
+        $process.Length      | Should -Be 1
+        $process.Id          | Should -BeGreaterThan 1
+    }
+
+    It "Should run without errors when -ArgumentList is empty" {
+        $process = Start-Process $pingCommand -ArgumentList $null -PassThru -WindowStyle Hidden
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Should it be `-ArgumentList "" ` ?

**Comment by markekraus** [comment]
@BrucePay can you please push an empty commit with `[feature]`? The tests in this PR are gated by the feature flag.

**Comment by SteveL-MSFT** [comment]
@BrucePay I believe GitHub won't auto-close issues in the title, so I put `Fix #4520` in the PR description which works.

For @markekraus' request, you can just do `git commit --amend` and update the description where the first line is `[feature]` which will kick off the feature tests in CI

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 2]
```diff
@@ -124,14 +124,20 @@ Describe "Start-Process" -Tags @("Feature") {
...
 
-    It "Should run without errors when -ArgumentList is empty" {
-        $process = Start-Process $pingCommand -ArgumentList $null -PassThru -WindowStyle Hidden
+    It "Should run without errors when -ArgumentList is @()" {
+        $process = Start-Process $pingCommand -ArgumentList @() -PassThru -WindowStyle Hidden
+        $process.Length      | Should -Be 1
+        $process.Id          | Should -BeGreaterThan 1
+    }
+
+    It "Should run without errors when -ArgumentList is ''" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: We could use `-TestCases` and replace last three tests with one.

**Comment by SteveL-MSFT** [comment]
@BrucePay looks like you accidentally pulled in other commits into this PR.  Find me tomorrow in the office to get this cleaned up.

**Comment by BrucePay** [comment]
@SteveL-MSFT Not sure how it happened bu oh my yes :-( I will definitely find you.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 (hunk 2 above)
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: I'd rather do this in a separate check-in. This whole file needs serious clean up e.g. there are tests that depend on notepad.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 3]
```diff
@@ -116,13 +120,31 @@ Describe "Start-Process" -Tags @("Feature") {
...
 
     It "Should return null when using -WhatIf switch with -PassThru" {
         Start-Process $pingCommand -ArgumentList $pingParam -PassThru -WhatIf | Should -BeNullOrEmpty
-   }
+    }
+
+    It 'Should run without errors when -ArgumentList is $null' {
+         $process = Start-Process $pingCommand -ArgumentList $null -PassThru @extraArgs
+         $process.Length      | Should -Be 1
+         $process.Id          | Should -BeGreaterThan 1
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Is this check required?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 4]
```diff
@@ -116,13 +120,31 @@ Describe "Start-Process" -Tags @("Feature") {
...
+    }
+ 
+    It "Should run without errors when -ArgumentList is @()" {
+        $process = Start-Process $pingCommand -ArgumentList @() -PassThru @extraArgs
+        $process.Length      | Should -Be 1
+        $process.Id          | Should -BeGreaterThan 1
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Same as above.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 5]
```diff
@@ -116,13 +120,31 @@ Describe "Start-Process" -Tags @("Feature") {
...
+    }
+
+    It "Should run without errors when -ArgumentList is ''" {
+        $process = Start-Process $pingCommand -ArgumentList '' -PassThru @extraArgs
+        $process.Length      | Should -Be 1
+        $process.Id          | Should -BeGreaterThan 1
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Same as above.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]
Please submit the last commit with `[Feature]` in the commit message to execute the tests in the PR.
```
git commit --allow-empty
git push
```

**Review by PaulHigin — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 (hunk 3 above)
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: These are basic sanity tests to make sure the process was created and is running (has an ID).

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 (hunk 4 above)
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: These are basic sanity tests to make sure the process was created and is running (has an ID).

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 (hunk 5 above)
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: These are basic sanity tests to make sure the process was created and is running (has an ID).

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
pushed a commit with `[Feature]` to kick off all tests.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 (hunk 3 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: If the process is not started the `$process` is `$null`.

**Comment by TravisEz13** [comment]
@adityapatwardhan Please make sure you check the PR title before merging.  The `Fix for #4520` should not be in the PR title.

**Comment by adityapatwardhan** [comment]
@TravisEz13 Yes agreed, missed it.

**Pull request merged**